add download tool aria2 support

bradfordzhang opened the pull request:
Because of some countries' network problem such as China, compilers can't download smoothly. If use aria2, it can train your own server download speed memory, and automatically select a mirror that is more suitable for you in the next crawl

dangowrt commented on scripts/download.pl:
```diff
@@ -73,16 +73,25 @@ ()
+
 	return $have_curl
-		? (qw(curl -f --connect-timeout 20 --retry 5 --location), $check_certificate ? '' : '--insecure', shellwords($ENV{CURL_OPTIONS} || ''), $url)
+		? ("aria2c --stderr $murl -c -x2 -s10 -j10 -k1M --server-stat-of=/dev/shm/spp --server-stat-if=/dev/shm/spp -d /dev/shm -o $rfn; cat /dev/shm/$rfn; rm /dev/shm/$rfn")
```
This looks wrong. Probably you should introduce another variable `have_aria2c` and decide based on that to use aria2c (if found), fall back to curl (if found) and finally fall back to wget.

SkyTeamWater replied:
hi dangowrt!
When I use 
```perl
my $have_aria2c = 0;
if (open ARIA2C, '-|', 'aria2c', '--version') {
	if (defined(my $line = readline ARIA2C)) {
		$have_aria2c = 1 if $line =~ /^aria2c /;
	}
	close ARIA2C;
}
```
variable `have_aria2c` always is `true`, I'm new to perl, I don't know what's going on.

SkyTeamWater replied:
And I need to judge if aria2c found at line 93, 153, 303 and 305. I tried to define `my $have_aria2c = 0;` at line 31, but when I `make download -j1 V=sc`, terminal tells me `Bareword "have_aria2c" not allowed while "strict subs" in use at /home/bradford/workspace/openwrt/scripts/download.pl line 167.`. What should I do?

dangowrt commented on scripts/download.pl:
```diff
@@ -140,7 +150,7 @@ sub download
-		my @cmd = download_cmd("$mirror/$download_filename");
+		my @cmd = download_cmd("$mirror/$download_filename", $download_filename, @mrs);
 		print STDERR "+ ".join(" ",@cmd)."\n";
 		open(FETCH_FD, '-|', @cmd) or die "Cannot launch curl or wget.\n";
```
This error message should be change to into `"Cannot launch aria2c, curl or wget.\n"`

dangowrt reviewed (changes requested):
Very nice idea! Please fix two minor issues.

Ansuel commented:
@SkyTeamWater ok idea but you have to rebase this and squash everything in one commit... also @dangowrt should we add aria2c as a dependency or keep this as an optional nice feature?

Also pls use correct title (something like `download.pl: ass aria2c support`) and provide a correct commit description

SkyTeamWater commented:
> @SkyTeamWater ok idea but you have to rebase this and squash everything in one commit... also @dangowrt should we add aria2c as a dependency or keep this as an optional nice feature?
> 
> Also pls use correct title (something like `download.pl: add aria2c support`) and provide a correct commit description

@Ansuel hi Ansuel, I've rebased my commits!

Ansuel commented on scripts/download.pl:
```diff
@@ -70,20 +70,41 @@ ()
+	}
+
+	print ("HAVE_ARIA2C:");
+	print ("$have_aria2c\n");
```
debug print?

SkyTeamWater replied:
so sry i forget to delete it

jow- commented:
> should we add aria2c as a dependency or keep this as an optiona

Please keep it optional!

Ansuel commented on scripts/download.pl:
```diff
@@ -70,20 +70,41 @@ ()
-	return $have_curl
-		? (qw(curl -f --connect-timeout 20 --retry 5 --location), $check_certificate ? '' : '--insecure', shellwords($ENV{CURL_OPTIONS} || ''), $url)
-		: (qw(wget --tries=5 --timeout=20 --output-document=-), $check_certificate ? '' : '--no-check-certificate', shellwords($ENV{WGET_OPTIONS} || ''), $url)
+	return $have_aria2c
```
Think here we should use a full `if() else` than `$have_aria2c ? :` just to make things more readable.

also AFAIK also aria2c supports disabling/enabling TLS verification. Can you add this here for consistency with what curl does?

SkyTeamWater replied:
sure i'll do that

SkyTeamWater commented:
There're so many variables, so that I think use `qw()` is not readable,so I just `if ($check_certificate) ... else ...`

SkyTeamWater commented:
so sorry for slow coding. my laptop runs so slow that i can't test `make download` smoothly

SkyTeamWater commented:
@Ansuel hi Christain,i've fixed the problem and rebased all commits into one.Can you review it?thanks

neheb commented:
Needs another rebase.

Ansuel commented:
@SkyTeamWater why closed :(

SkyTeamWater commented:
> @SkyTeamWater why closed :(

i rebased agai. wait a minute

bradfordzhang committed:
download.pl: add aria2c support

Use aria2c download tool by default on package download if available in
the system.
aria2c permits to use multiple mirrors and may improve download speed on
special context where servers are hard to reach.

Co-authored-by: Christian Marangi <ansuelsmth@gmail.com>
Signed-off-by: Bradford Zhang <zyc@zyc.name>
[ fix wrong var in the script and improve commit description ]
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

SkyTeamWater commented:
> @SkyTeamWater why closed :(

what should i do now?

nasbdh9 commented:
git pull --rebase

SkyTeamWater commented:
> git pull --rebase

Already up to date.
😂

nasbdh9 commented:
git reset --hard cf2c9498bed5e2f34170ada394be26348949f107
git cherry-pick 2b039e305e007c0174a6bb93fa0be9601d176a33
git push origin aria2-download -f

SkyTeamWater commented:
> git reset --hard cf2c9498bed5e2f34170ada394be26348949f107
> git cherry-pick 2b039e305e007c0174a6bb93fa0be9601d176a33
> git push origin aria2-download -f

I'm done. Sooooo thank u

Ansuel commented on scripts/download.pl:
```diff
@@ -70,27 +70,44 @@ ()
+			return ("aria2c --stderr $murl -c -x2 -s10 -j10 -k1M --server-stat-of=/dev/shm/spp --server-stat-if=/dev/shm/spp -d /dev/shm -o $rfn --check-certificate=false; cat /dev/shm/$rfn; rm /dev/shm/$rfn");
+		}
+	} elsif ($have_curl) {
+		return (qw(curl -f --connect-timeout 20 --retry 5 --location), $check_certificate ? '' : '--insecure', shellwords($ENV{CURL_OPTIONS} || ''), $url);
```
```suggestion
		return (qw(curl -f --connect-timeout 20 --retry 5 --location),
			$check_certificate ? () : '--insecure',
			shellwords($ENV{CURL_OPTIONS} || ''),
			$url);
```

Ansuel commented on scripts/download.pl:
```diff
@@ -70,27 +70,44 @@ ()
+	} elsif ($have_curl) {
+		return (qw(curl -f --connect-timeout 20 --retry 5 --location), $check_certificate ? '' : '--insecure', shellwords($ENV{CURL_OPTIONS} || ''), $url);
+	} else {
+		return (qw(wget --tries=5 --timeout=20 --output-document=-), $check_certificate ? '' : '--no-check-certificate', shellwords($ENV{WGET_OPTIONS} || ''), $url);
```
```suggestion
		return (qw(wget --tries=5 --timeout=20 --output-document=-),
			$check_certificate ? () : '--no-check-certificate',
			shellwords($ENV{WGET_OPTIONS} || ''),
			$url);
```

bradfordzhang commented:
@Ansuel Hi Christain, can we merge now?

Ansuel commented:
@bradfordzhang sorry i'm reworking the aria2c wait a second so i can test and suggest and this can merged :D

bradfordzhang commented:
> 

Copy that.Thank you

Ansuel commented on scripts/download.pl:
```diff
@@ -70,27 +70,50 @@ ()
-	return $have_curl
-		? (qw(curl -f --connect-timeout 20 --retry 5 --location),
+	if ($have_aria2c) {
+		if ($check_certificate) {
+			return ("aria2c --stderr $murl -c -x2 -s10 -j10 -k1M --server-stat-of=/dev/shm/spp --server-stat-if=/dev/shm/spp -d /dev/shm -o $rfn; cat /dev/shm/$rfn; rm /dev/shm/$rfn");
+		} else {
+			return ("aria2c --stderr $murl -c -x2 -s10 -j10 -k1M --server-stat-of=/dev/shm/spp --server-stat-if=/dev/shm/spp -d /dev/shm -o $rfn --check-certificate=false; cat /dev/shm/$rfn; rm /dev/shm/$rfn");
+		}
```
```suggestion
		return join(" ", "touch /dev/shm/${rfn}_spp;",
			qw(aria2c --stderr -c -x2 -s10 -j10 -k1M), $murl,
			$check_certificate ? () : '--check-certificate=false',
			"--server-stat-of=/dev/shm/${rfn}_spp",
			"--server-stat-if=/dev/shm/${rfn}_spp",
			"-d /dev/shm -o $rfn;",
			"cat /dev/shm/$rfn;", "rm /dev/shm/$rfn /dev/shm/${rfn}_spp");
```

What do you think of this solution?

Also I notice that in your implementation you didn't create the ssp in the first place (so it wasn't actually used) and also it's used by all the packages. Is that intended?

Ansuel commented on scripts/download.pl:
```diff
@@ -100,6 +123,7 @@ sub download
 {
 	my $mirror = shift;
 	my $download_filename = shift;
+	my @mrs = @_;
```
lets make this more readable.
```suggestion
	my @additional_mirrors= @_;
```

Ansuel commented on scripts/download.pl:
```diff
@@ -146,9 +170,9 @@ sub download
 		};
 	} else {
-		my @cmd = download_cmd("$mirror/$download_filename");
+		my @cmd = download_cmd("$mirror/$download_filename", $download_filename, @mrs);
```
```suggestion
		my @cmd = download_cmd("$mirror/$download_filename", $download_filename, @additional_mirrors);
```

Ansuel commented on scripts/download.pl:
```diff
@@ -296,10 +320,11 @@ sub cleanup
 	$mirror or die "No more mirrors to try - giving up.\n";
 
-	download($mirror, $url_filename);
+	download($mirror, $url_filename, @mirrors);
```
```suggestion
	download($mirror, $url_filename, @additional_mirrors);
```

Ansuel commented on scripts/download.pl:
```diff
@@ -296,10 +320,11 @@ sub cleanup
+	download($mirror, $url_filename, @mirrors);
 	if (!-f "$target/$filename" && $url_filename ne $filename) {
-		download($mirror, $filename);
+		download($mirror, $filename, @mirrors);
```
```suggestion
		download($mirror, $filename, @additional_mirrors);
```

Ansuel commented on scripts/download.pl:
```diff
@@ -70,27 +70,50 @@ ()
-
-	if (open CURL, "curl --version 2>/dev/null |") {
+	my $have_aria2c = 0;
+	my $fn = shift;
```
in theory we should be able to drop this and use directly $filename as it's global

Ansuel commented on scripts/download.pl:
```diff
@@ -70,27 +70,50 @@ ()
-	if (open CURL, "curl --version 2>/dev/null |") {
+	my $have_aria2c = 0;
+	my $fn = shift;
+	my @mrs = @_;
```
```suggestion
	my @additional_mirrors = @_;
```

Ansuel replied:
Also if we really want we can reduce this to `my $additional_mirrors =join(" ", map "$_/$filename", @_);` and drop the for later... 

And in the aria2c download function just use

`qw(aria2c --stderr -c -x2 -s10 -j10 -k1M), $url, $additional_mirrors ,`

Ansuel commented on scripts/download.pl:
```diff
@@ -70,27 +70,50 @@ ()
+		close ARIA2C;
+	}
+
+	for my $el (@mrs) {
```
```suggestion
	for my $mirror (@additional_mirrors ) {
```

Ansuel commented on scripts/download.pl:
```diff
@@ -70,27 +70,50 @@ ()
+	}
+
+	for my $el (@mrs) {
+		$murl = $murl." '$el/$fn'";
```
```suggestion
		$mirrors_url = $mirrors_url ." '$mirror /$filename'";
```

Ansuel commented on scripts/download.pl:
```diff
@@ -70,27 +70,50 @@ ()
+	my $have_aria2c = 0;
+	my $fn = shift;
+	my @mrs = @_;
+	my $murl = "'$url'";
```
```suggestion
	my $mirrors_url = "'$url'";
```

Ansuel commented on scripts/download.pl:
```diff
@@ -70,27 +70,50 @@ ()
+	my $murl = "'$url'";
+
+	my @chArray = ('a'..'z', 'A'..'Z', 0..9);
+	my $rfn = join '', map{ $chArray[int rand @chArray] } 0..9;
```
Should we append the filename to the random string? Also wonder if a better naming would be filename_token or something like that?

bradfordzhang replied:
I think random string is enough

Ansuel commented on scripts/download.pl:
```diff
@@ -296,10 +322,11 @@ sub cleanup
+	download($mirror, $url_filename, @additional_mirrors);
 	if (!-f "$target/$filename" && $url_filename ne $filename) {
-		download($mirror, $filename);
+		download($mirror, $filename, @additional_mirrors);
```
This and the other cause aria2 not working. Pls revert in this function, additional_mirrors to mirrors.

With that fixed, I will merge this :D

Ansuel commented:
@bradfordzhang Hi, the script can still be improved but I fixed one last error, improved the commit description and I decided to merge so you can use this feature :D

bradfordzhang commented:
> @bradfordzhang Hi, the script can still be improved but I fixed one last error, improved the commit description and I decided to merge so you can use this feature :D

so thank you about that

feckert commented:
@Ansuel 
I do not have aria2 installed an my system. So the download.pl uses wget or curl.

I think we have a problem now with this change that it will not try new download servers if the package is not in the dl folder.
If I am using aria2 it works as expected, but without aria2 I have problems to download new package version which are not in my dl folder. 
```
cp: cannot stat '/home/feckert/dl/stunnel-5.66.tar.gz': No such file or directory
mkdir -p /home/feckert/dl
SHELL= flock /home/feckert/workspace/openwrt/LDM-master-x86_64/build/openwrt/tmp/.stunnel-5.66.tar.gz.flock -c '        /home/feckert/workspace/openwrt/LDM-master-x86_64/build/openwrt/scripts/download.pl "/home/feckert/dl" "stunnel-5.66.tar.gz" "558178704d1aa5f6883aac6cc5d6bbf2a5714c8a0d2e91da0392468cee9f579c" "" "http://ftp.nluug.nl/pub/networking/stunnel/" "http://www.usenix.org.uk/mirrors/stunnel/" "https://www.stunnel.org/downloads/" "https://www.usenix.org.uk/mirrors/stunnel/archive/5.x/"    '
Can't exec "aria2c": No such file or directory at /home/feckert/workspace/openwrt/LDM-master-x86_64/build/openwrt/scripts/download.pl line 88.
+ curl -f --connect-timeout 20 --retry 5 --location ftp://anonymous:anonymous@nas.dev.tdt.de/openwrt/sources/stunnel-5.66.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (78) The file does not exist
Download failed.
```

Ansuel commented:
The problem seems to be related mirrors getting removed. I have a pr open.
Can you test if that fix it?

Il Lun 26 Set 2022, 16:33 Florian Eckert ***@***.***> ha
scritto:

> @Ansuel <https://github.com/Ansuel>
> I do not have aria2 installed an my system. So the download.pl uses wget
> or curl.
>
> I think we have a problem now with this change that it will not try new
> download servers if the package is not in the dl folder.
> If I am using aria2 it works as expected, but without aria2 I have
> problems to download new package version which are not in my dl folder.
>
> cp: cannot stat '/home/feckert/dl/stunnel-5.66.tar.gz': No such file or directory
> mkdir -p /home/feckert/dl
> SHELL= flock /home/feckert/workspace/openwrt/LDM-master-x86_64/build/openwrt/tmp/.stunnel-5.66.tar.gz.flock -c '        /home/feckert/workspace/openwrt/LDM-master-x86_64/build/openwrt/scripts/download.pl "/home/feckert/dl" "stunnel-5.66.tar.gz" "558178704d1aa5f6883aac6cc5d6bbf2a5714c8a0d2e91da0392468cee9f579c" "" "http://ftp.nluug.nl/pub/networking/stunnel/" "http://www.usenix.org.uk/mirrors/stunnel/" "https://www.stunnel.org/downloads/" "https://www.usenix.org.uk/mirrors/stunnel/archive/5.x/"    '
> Can't exec "aria2c": No such file or directory at /home/feckert/workspace/openwrt/LDM-master-x86_64/build/openwrt/scripts/download.pl line 88.
> + curl -f --connect-timeout 20 --retry 5 --location ***@***.***/openwrt/sources/stunnel-5.66.tar.gz
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> curl: (78) The file does not exist
> Download failed.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/10682#issuecomment-1258133797>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQT4EAZ7DI6YI35AS4TWAGX3DANCNFSM6AAAAAAQJ25OBM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

feckert commented:
@Ansuel Sorry for the late repley. The latest changes fixes my issue. Thanks
